src/main/scala/com/kinja/amqp/AmqpConsumer.scala

```diff
@@ -61,69 +76,120 @@ class AmqpConsumer(
+	 * than this number.
+	 */
+	private val toleranceNanos = 10000000L // 10 milliseconds
+
+	/**
+	 * Default state of the listener which receives messages from RabbitMQ.
+	 */
+	def receive = {
+		case Delivery(consumerTag, envelope, properties, body) =>
+			val nextTickNanos = System.nanoTime + spacing.toNanos
+			val s = new String(body, "UTF-8")
+
+			implicitly[Reads[A]].reads(s) match {
+				case Right(message) =>
+					try {
+						Await.result(processor(message), timeout)
+						val ack = Ack(envelope.getDeliveryTag)
+
+						// sleep until we are allowd to receive a new message
+						val nowNanos = System.nanoTime
+						if (nowNanos < nextTickNanos - toleranceNanos) {
+							implicit val ec: ExecutionContext = context.dispatcher
+							context.system.scheduler.scheduleOnce((nextTickNanos - nowNanos).nanos, self, WakeUp)
+							context.become(asleep(sender, ack))
+						} else {
+							sender ! ack
+						}
+					} catch {
+						case NonFatal(t) =>
+							logger.warn(s"""[RabbitMQ] Exception while processing message "$s" : $t""")
+							sender ! Reject(envelope.getDeliveryTag, requeue = true)
 					}
+				case Left(e) =>
+					logger.warn(s"""[RabbitMQ] Couldn't parse message "$s" : $e""")
+					sender ! Reject(envelope.getDeliveryTag, requeue = false)
 			}
+		case WakeUp => ()
+	}
 
-			def asleep(originalSender: ActorRef, ack: Ack): Receive = {
-				case WakeUp =>
-					originalSender ! ack
-					context.unbecome()
-				case Delivery(_, envelope, _, _) =>
-					sender ! Reject(envelope.getDeliveryTag, requeue = true)
-			}
-		}))
+	def asleep(originalSender: ActorRef, ack: Ack): Receive = {
+		case WakeUp =>
+			originalSender ! ack
+			context.unbecome()
+		case Delivery(_, envelope, _, _) =>
+			sender ! Reject(envelope.getDeliveryTag, requeue = true)
 	}
 }
+
+class Rejecter extends Actor {
+	def receive = {
+		case Delivery(_, envelope, _, _) =>
+			sender ! Reject(envelope.getDeliveryTag, requeue = true)
+	}
+}
+
+final case class SetListener(listener: ActorRef)
+
+class Proxy extends Actor {
+	def receive = {
+		case SetListener(listener) => context.become(forwarding(listener))
+	}
+	def forwarding(listener: ActorRef): Receive = {
+		case SetListener(otherListener) => context.become(forwarding(otherListener))
+		case message => listener.forward(message)
+	}
+}
+
+case object Connect
+case object Disconnect
+
+class Supervisor[A: Reads](
+	timeout: FiniteDuration,
+	spacing: FiniteDuration,
+	processor: A => Future[Unit],
+	rejecter: ActorRef,
+	proxy: ActorRef,
+	logger: Slf4jLogger) extends Actor {
+
+	def receive = {
+		case Disconnect => ()
+		case Connect =>
+			val listener = context.actorOf(Props(new Listener(timeout, spacing, processor, logger)))
+			proxy ! SetListener(listener)
+			context.become(connected(listener))
+
+	}
+
+	def connected(listener: ActorRef): Receive = {
+		case Connect => ()
+		case Disconnect =>
+			context.stop(listener) // Let listener handle one more message
```

Comment:
According to [StackOverflow](https://stackoverflow.com/questions/13847963/akka-kill-vs-stop-vs-poison-pill), the message being processed is completed, and the rest of the queue is then sent to the dead letters mailbox. I'm wondering if there's a way to capture those from the dead letter mailbox and actively reject them with RabbitMQ instead of just waiting for RabbitMQ to treat them as lost messages.

Reply:
According to ![Akka docs](https://doc.akka.io/docs/akka/current/scala/event-bus.html#dead-letters), dead letters mailbox is supposed to be a debug feature, and lost messages are not guaranteed to arrive there.

Reply:
Sorry, wrong link: ![here](https://doc.akka.io/docs/akka/current/scala/general/message-delivery-reliability.html#what-should-i-use-dead-letters-for-)

Reply:
Sorry, wrong link: [here](https://doc.akka.io/docs/akka/current/scala/general/message-delivery-reliability.html#what-should-i-use-dead-letters-for-)